File: judge/history.mjs
// Independent synthetic history oracle. Request expectations are authored from
// the scenario, never inferred from successful product output or its cache.
export function historyScenarios({ runNode, runCli, check, fixture, initPath, matrixPath, invokeCli, cliBin, root, spawnSync, resolve, readFileSync, writeFileSync }) {
  const codes = ["10", "20", "30", "40", "50", "60", "70", "80"];
  const tenors = ["3M", "6M", "9M", "1Y", "1.5Y", "2Y", "2.5Y", "3Y", "5Y", "7Y", "10Y", "15Y", "20Y", "30Y", "50Y"];
  const keys = ["m3", "m6", "m9", "y1", "y15a", "y2", "y25", "y3", "y5", "y7", "y10", "y15", "y20", "y30", "y50"];
  const xml = rows => `<Root xmlns="http://www.nexacroplatform.com/platform/dataset"><Parameters><Parameter id="ErrorCode">0</Parameter></Parameters><Dataset id="output1"><Rows>${rows}</Rows></Dataset></Root>`;
  const catalog = extra => xml(`<Row><Col id="divCode">10</Col><Col id="divName">국채</Col></Row>${extra ? `<Row><Col id="divCode">${extra}</Col><Col id="divName">추가 ${extra}</Col></Row>` : ""}`);
  const matrix = xml(`<Row><Col id="pricingGroupCode">001</Col><Col id="pricingGroupName">=한글</Col>${keys.map((key, index) => `<Col id="${key}">${index === 0 ? "-0.500" : index === 1 ? "0.000" : "-"}</Col>`).join("")}</Row>`);
  const init = (date, body = catalog()) => ({ path: initPath, body, expectedCells: { calBaseDt: date } });
  const pair = (date, code, body = matrix) => ({ path: matrixPath, body, expectedCells: { calBaseDt: date, cboYtmSort: code } });
  const day = (date, extra) => [init(date, catalog(extra)), ...[...codes, ...(extra ? [extra] : [])].map(code => pair(date, code))];
  const cases = [
    { name: "dynamic-catalog", input: { baseDates: ["20260609", "2026.06.08", "2026-06-09"] }, steps: [...day("20260608", "090"), ...day("20260609", "091")], available: 18, unavailable: 0, dates: ["2026-06-08", "2026-06-09"] },
    { name: "partial", input: { startDate: "2026-06-08", endDate: "2026-06-09" }, steps: [...day("20260608"), init("20260609"), ...codes.map(code => pair("20260609", code, code === "80" ? xml("") : matrix))], available: 15, unavailable: 1, dates: ["2026-06-08", "2026-06-09"] },
    { name: "discovery-unavailable", input: { baseDates: ["20260608", "20260609"] }, steps: [init("20260608", xml("")), init("20260609", xml(""))], available: 0, unavailable: 16, dates: ["2026-06-08", "2026-06-09"] },
    { name: "fallback-reuse", input: { baseDates: ["20260608", "20260609"], fallback: "previous-available", lookbackDays: 2 }, steps: [...day("20260608"), init("20260609"), ...codes.map(code => pair("20260609", code, xml("")))], available: 16, unavailable: 0, dates: ["2026-06-08", "2026-06-09"] },
    { name: "fallback-discovery", input: { baseDates: ["20260609"], fallback: "previous-available", lookbackDays: 2 }, steps: [init("20260609", xml("")), ...day("20260608")], available: 8, unavailable: 0, dates: ["2026-06-09"] },
    { name: "fallback-exhausted", input: { baseDates: ["20260609"], fallback: "previous-available", lookbackDays: 1 }, steps: [init("20260609", xml("")), init("20260608", xml(""))], available: 0, unavailable: 8, dates: ["2026-06-09"] },
    { name: "leap-range", input: { startDate: "2024-02-28", endDate: "2024-03-01" }, steps: ["20240228", "20240229", "20240301"].flatMap(date => day(date)), available: 24, unavailable: 0, dates: ["2024-02-28", "2024-02-29", "2024-03-01"] }
  ];
  function args(input) {
    return ["history", ...(input.baseDates ? input.baseDates.flatMap(date => ["--base-date", date]) : ["--start-date", input.startDate, "--end-date", input.endDate]), ...(input.fallback ? ["--fallback", input.fallback, "--lookback-days", String(input.lookbackDays)] : [])];
  }
  function resultCheck(value, c, label) {
    check(value?.availableCount === c.available && value?.unavailableCount === c.unavailable && value?.dataRowCount === c.available, `${label} pair/data counts`);
    check(JSON.stringify(value?.requestedDates) === JSON.stringify(c.dates), `${label} normalized ordered dates`);
    check(value?.entries?.length === c.available + c.unavailable, `${label} every pair represented once`);
    for (const date of c.dates) {
      const entries = value?.entries?.filter(e => (e.matrix?.requestedBaseDate || e.requestedBaseDate) === date) || [];
      check(codes.every(code => entries.some(e => (e.matrix?.kind || e.kind)?.code === code)), `${label} canonical omissions retained including 80`);
    }
    for (const entry of value?.entries || []) {
      if (entry.availability === "unavailable") {
        check(!entry.matrix && entry.reason.includes("source_data_unavailable"), `${label} missingness is tagged without observations`);
        continue;
      }
      const m = entry.matrix;
      check(m.rows.length === 1 && m.rows[0].pricingGroupCode === "001" && m.rows[0].pricingGroupName === "=한글", `${label} source identity`);
      check(m.rows[0].yields["3M"] === -0.5 && m.rows[0].yields["6M"] === 0 && m.rows[0].yields["9M"] === null && m.rows[0].yieldText["3M"] === "-0.500", `${label} numeric and raw values`);
      if (c.name.startsWith("fallback")) check(m.baseDate === "2026-06-08", `${label} actual date preserved`);
    }
    if (c.name === "dynamic-catalog") check(value.entries.map(e => e.matrix.kind.code).join(",") === [...codes,"090",...codes,"091"].join(","), `${label} live-only membership stays date-local`);
    if (c.name === "fallback-discovery") check(value.discovery[0].available === false, `${label} fallback does not relabel discovery`);
    if (c.name === "fallback-reuse") check(value.entries.slice(8).every(e => e.matrix.dateResolution.attemptedDates.join(",") === "2026-06-09,2026-06-08"), `${label} cache retains logical attempts`);
  }
  for (const c of cases) {
    const config = fixture(c.steps);
    runNode(`history:${c.name}`, { action: "execute", operation: "history", input: c.input }, config, (result, label) => { check(result.ok, `${label} succeeds`); if (result.ok) resultCheck(result.value, c, label); });
    runCli(`history:${c.name}`, [...args(c.input), "--format=json"], config, (result, label) => { check(result.status === 0, `${label} succeeds`); if (result.status === 0) resultCheck(JSON.parse(result.stdout).result, c, label); }, { isolated: true, requestPayload: { operation: "history", input: c.input } });
    for (const format of ["csv", "tsv"]) runCli(`history:${c.name}:${format}`, [...args(c.input), `--format=${format}`], config, (result,label) => {
      check(result.status === 0 && result.stdout.split("\n").length === c.available + c.unavailable + 2, `${label} text retains unavailable pairs`);
      check(result.stdout.includes("availability") && (c.unavailable === 0 || result.stdout.includes("unavailable")), `${label} explicit availability`);
    });
    runCli(`history:${c.name}:xlsx`, cwd => [...args(c.input), "--format=xlsx", `--output=${resolve(cwd,"history.xlsx")}`], config, (result, label, cwd) => {
      check(result.status === 0, `${label} export succeeds`);
      if (result.status !== 0) return;
      const receipt = JSON.parse(result.stdout).result;
      check(receipt.rowCount === c.available && receipt.availableCount === c.available && receipt.unavailableCount === c.unavailable && receipt.dataRowCount === c.available, `${label} complete receipt`);
      const reference = JSON.parse(invokeCli(cliBin, [...args(c.input), "--format=json"], config).stdout).result;
      const inspected = spawnSync(process.env.PYO3_PYTHON || "python3", [resolve(root,"judge/inspect-xlsx.py"), resolve(cwd,"history.xlsx")], { encoding:"utf8", timeout:15000, maxBuffer:8*1024*1024 });
      check(inspected.status === 0, `${label} independent OOXML parser: ${inspected.stderr}`);
      if (inspected.status === 0) workbook(JSON.parse(inspected.stdout), reference, label);
      check(readFileSync(resolve(cwd,"history.xlsx")).subarray(0,2).toString() === "PK", `${label} published ZIP`);
    }, { isolated:true, golden:false, requestPayload:{ operation:"history", input:c.input } });
  }
  function workbook(w, expected, label) {
    check(w.sheets.map(s=>s.name).join(",") === "History,Availability,Metadata", `${label} exactly three sheets`);
    const [data, availability, metadata] = w.sheets;
    if (!data || !availability || !metadata) return;
    const col = i => i < 26 ? String.fromCharCode(65+i) : `A${String.fromCharCode(65+i-26)}`;
    const cell = (sheet,r,c,value) => {
      const actual = sheet.cells[`${col(c)}${r}`];
      const type = value === null ? "empty" : typeof value === "string" ? "text" : typeof value;
      check(actual?.type === type && actual?.value === value, `${label} ${sheet.name}!${col(c)}${r} typed value`);
      if (sheet === data && type === "number") check(actual.format === "0.000", `${label} unscaled yield format`);
    };
    const columns = ["requestedBaseDate","baseDate","usedFallback","kindCode","kindName","pricingGroupCode","pricingGroupName",...tenors];
    columns.forEach((v,c)=>cell(data,1,c,v));
    let r=2;
    for (const e of expected.entries) if (e.availability === "available") for (const row of e.matrix.rows) {
      const m=e.matrix;
      [m.requestedBaseDate,m.baseDate,m.dateResolution.usedFallback,m.kind.code,m.kind.name,row.pricingGroupCode,row.pricingGroupName,...tenors.map(t=>row.yields[t])].forEach((v,c)=>cell(data,r,c,v)); r++;
    }
    check(Object.keys(data.cells).length === columns.length*r - columns.length, `${label} exact data dimensions`);
    check(data.pane?.xSplit === "7" && data.pane.ySplit === "1" && data.pane.state === "frozen", `${label} freeze identities and header`);
    check(data.filter === `A1:V${r-1}`, `${label} data filter`);
    expected.entries.forEach((e,index) => {
      const m=e.matrix, requested=m?.requestedBaseDate || e.requestedBaseDate, kind=m?.kind || e.kind;
      [requested,kind.code,kind.name,e.availability,m?.baseDate || null,m ? m.dateResolution.usedFallback : null,m?.rows.length || 0,e.reason || null,expected.discovery.find(d=>d.requestedBaseDate===requested).available].forEach((v,c)=>cell(availability,index+2,c,v));
    });
    check(availability.filter === `A1:I${expected.entries.length+1}`, `${label} complete availability filter`);
    const fields={};
    for(let i=2;metadata.cells[`A${i}`];i++) fields[metadata.cells[`A${i}`].value]=metadata.cells[`B${i}`]?.value;
    check(fields.operation === "history" && fields.availableCount === String(expected.availableCount) && fields.unavailableCount === String(expected.unavailableCount), `${label} metadata counts`);
    expected.discovery.forEach((d,i)=>check(fields[`discovery[${i}].available`]===String(d.available), `${label} discovery provenance`));
    expected.entries.forEach((e,i)=> {
      const m=e.matrix;
      (m?.dateResolution.attemptedDates || e.attemptedDates).forEach((date,j)=>check(fields[`entries[${i}].attemptedDates[${j}]`] === date, `${label} attempted dates`));
      if(m) check(fields[`entries[${i}].source.request.parameters.cboYtmSort`] === m.kind.code, `${label} source code provenance`);
    });
  }
  for (const input of [{}, {baseDates:[]}, {baseDates:["2026-02-30"]}, {baseDates:["20260608"],startDate:"2026-06-08",endDate:"2026-06-09"}, {startDate:"2026-06-09",endDate:"2026-06-08"}, {startDate:"2020-01-01",endDate:"2026-01-01"}, {baseDates:["20260608"],lookbackDays:1}]) {
    runNode(`history:invalid-${JSON.stringify(input)}`, {action:"execute",operation:"history",input}, fixture([]), (r,label)=>check(!r.ok && r.error?.code === "invalid_parameter",`${label} invalid before source`));
  }
  runNode("history:sparse-input", {action:"history-sparse-input"}, fixture([]), (r,label)=>check(!r.ok && r.error?.code === "invalid_parameter", `${label} rejects holes before native serialization`));
  const fatal = fixture([init("20260608"), pair("20260608","10"), {...pair("20260608","20"), body:undefined, transportError:"synthetic failure"}]);
  runNode("history:fatal",{action:"execute",operation:"history",input:{baseDates:["20260608","20260609"]}},fatal,(r,label)=>check(!r.ok && r.error?.actual?.kind?.code === "20" && r.error?.operationName === "history",`${label} fatal pair context`));
  runCli("history:fatal-preserves-file", ["history","--base-date","20260608","--format=xlsx","--output=history.xlsx","--overwrite"], fatal, (r,label,cwd)=> {
    check(r.status !== 0 && JSON.parse(r.stdout).error.actual.kind.code === "20", `${label} reports fatal pair`);
    check(readFileSync(resolve(cwd,"history.xlsx"),"utf8") === "prior workbook", `${label} preserves destination`);
  }, { isolated:true, requestPayload:{operation:"history",input:{baseDates:["20260608"]}}, setup:cwd=>writeFileSync(resolve(cwd,"history.xlsx"),"prior workbook") });
  runCli("history:preflight", ["history","--base-date","20260608","--format=xlsx","--output=history.xlsx"], fixture([]), (r,label)=>check(r.status !== 0 && JSON.parse(r.stdout).error.code === "output_exists", `${label} destination validation precedes requests`), {isolated:true,setup:cwd=>writeFileSync(resolve(cwd,"history.xlsx"),"prior workbook")});
  runNode("history:abort",{action:"abort-handler-preservation",operation:"history",input:{baseDates:["20260608"]}},fixture([{path:initPath,waitForCancellation:true,expectedCells:{calBaseDt:"20260608"}}]),(r,label)=>check(r.ok && r.value?.cancellationCode === "source_transport_error",`${label} abort reaches history`));
  runCli("history:help",["history","--help"],fixture([]),(r,label)=>check(r.status===0 && r.stdout.includes("2000") && r.stdout.includes("Availability"),`${label} offline help`));
}

File: crates/ytm-core/tests/history.rs
use async_trait::async_trait;
use std::{
    collections::HashSet,
    sync::{Arc, Mutex},
};
use ytm_core::{
    BaseDate, CancellationToken, DateSelection, FallbackPolicy, HistoryEntry, HistoryInput,
    HistoryRequest, LookbackDays, PreparedRequest, Transport, YtmClient, YtmError,
};

fn xml(rows: &str) -> Vec<u8> {
    format!(r#"<Root xmlns="http://www.nexacroplatform.com/platform/dataset"><Parameters><Parameter id="ErrorCode">0</Parameter></Parameters><Dataset id="output1"><Rows>{rows}</Rows></Dataset></Root>"#).into_bytes()
}
fn matrix() -> Vec<u8> {
    let mut row = String::from(
        r#"<Row><Col id="pricingGroupCode">001</Col><Col id="pricingGroupName">=한글</Col>"#,
    );
    for (index, key) in [
        "m3", "m6", "m9", "y1", "y15a", "y2", "y25", "y3", "y5", "y7", "y10", "y15", "y20", "y30",
        "y50",
    ]
    .iter()
    .enumerate()
    {
        row.push_str(&format!(
            r#"<Col id="{key}">{}</Col>"#,
            if index == 0 {
                "-0.500"
            } else if index == 1 {
                "0"
            } else {
                "-"
            }
        ));
    }
    row.push_str("</Row>");
    xml(&row)
}
fn cell(body: &str, key: &str) -> String {
    body.split(&format!("<Col id=\"{key}\"> ").replace("> ", ">"))
        .nth(1)
        .unwrap()
        .split("</Col>")
        .next()
        .unwrap()
        .to_owned()
}
#[derive(Clone, Default)]
struct Synthetic {
    requests: Arc<Mutex<Vec<(String, String, String)>>>,
    empty_discovery: bool,
    fallback: bool,
    fatal: bool,
    wait: bool,
}
#[async_trait]
impl Transport for Synthetic {
    async fn post(
        &self,
        request: PreparedRequest,
        cancellation: CancellationToken,
    ) -> Result<Vec<u8>, YtmError> {
        let date = cell(&request.body, "calBaseDt");
        let kind = cell(&request.body, "cboYtmSort");
        self.requests
            .lock()
            .unwrap()
            .push((request.operation.into(), date.clone(), kind.clone()));
        if self.wait {
            cancellation.cancelled().await;
            return Err(YtmError::transport("cancelled", None, Some("AbortError")));
        }
        if request.operation == "initializeYtmMatrix" {
            if self.empty_discovery {
                return Ok(xml(""));
            }
            let label = if date == "20260608" {
                "이전"
            } else {
                "이후"
            };
            // Deliberately omit seven canonical categories, including 80.
            return Ok(xml(&format!(
                r#"<Row><Col id="divCode">10</Col><Col id="divName">국채</Col></Row><Row><Col id="divCode">90</Col><Col id="divName">{label}</Col></Row>"#
            )));
        }
        if self.fatal && kind == "20" {
            return Err(YtmError::protocol(
                "17".into(),
                Some("synthetic failure".into()),
            ));
        }
        if self.fallback && date != "20260608" {
            return Ok(xml(""));
        }
        Ok(matrix())
    }
}
fn input(dates: &[&str], fallback: bool) -> HistoryInput {
    let mut input = HistoryInput::new(
        DateSelection::dates(dates.iter().map(|d| d.parse().unwrap()).collect()).unwrap(),
    );
    if fallback {
        input.fallback = FallbackPolicy::PreviousAvailable(LookbackDays::new(2).unwrap());
    }
    input
}
#[test]
fn date_selection_contract() {
    let selection = DateSelection::dates(vec![
        "20260609".parse().unwrap(),
        "2026.06.08".parse().unwrap(),
        "2026-06-08".parse().unwrap(),
    ])
    .unwrap();
    assert_eq!(
        selection
            .as_dates()
            .iter()
            .map(ToString::to_string)
            .collect::<Vec<_>>(),
        ["2026-06-08", "2026-06-09"]
    );
    assert_eq!(
        DateSelection::range("2024-02-28".parse().unwrap(), "2024-03-01".parse().unwrap())
            .unwrap()
            .as_dates()
            .len(),
        3
    );
    assert!(
        DateSelection::range("2024-03-01".parse().unwrap(), "2024-02-28".parse().unwrap()).is_err()
    );
    assert!(DateSelection::dates(vec![]).is_err());
    assert!(DateSelection::dates(vec!["2026-06-08".parse().unwrap(); 2001]).is_err());
    for value in [
        serde_json::json!({}),
        serde_json::json!({"baseDates":[],"startDate":"2024-01-01","endDate":"2024-01-01"}),
        serde_json::json!({"startDate":"2020-01-01","endDate":"2026-01-01"}),
        serde_json::json!({"baseDates":["2026-02-30"]}),
    ] {
        let request: HistoryRequest = serde_json::from_value(value).unwrap();
        assert!(HistoryInput::try_from(request).is_err());
    }
}
#[tokio::test]
async fn all_categories_reuse_and_observation_labels() {
    let transport = Synthetic {
        fallback: true,
        ..Default::default()
    };
    let client = YtmClient::with_transport(transport.clone());
    let result = client
        .history(input(&["20260609", "20260608", "2026.06.09"], true))
        .await
        .unwrap();
    assert_eq!(
        (
            result.available_count,
            result.unavailable_count,
            result.data_row_count
        ),
        (18, 0, 18)
    );
    assert_eq!(result.entries.len(), 18);
    let requests = transport.requests.lock().unwrap().clone();
    assert_eq!(requests.len(), 20); // Two catalogs + nine matrices per date.
    assert_eq!(
        requests.iter().collect::<HashSet<_>>().len(),
        requests.len()
    );
    for (index, entry) in result.entries.iter().enumerate() {
        let HistoryEntry::Available { matrix } = entry else {
            panic!("available")
        };
        assert_eq!(matrix.base_date.to_string(), "2026-06-08");
        assert_eq!(
            matrix.date_resolution.attempted_dates.len(),
            if index < 9 { 1 } else { 2 }
        );
        assert_eq!(matrix.rows[0].pricing_group_code, "001");
        assert_eq!(matrix.rows[0].yields["3M"], Some(-0.5));
        assert_eq!(matrix.rows[0].yields["6M"], Some(0.0));
        assert_eq!(matrix.rows[0].yields["9M"], None);
        assert_eq!(matrix.rows[0].group_name, matrix.kind.name);
        if matrix.kind.code == "90" {
            assert_eq!(matrix.kind.name, "이전");
        }
    }
    let alone = client.history(input(&["20260609"], true)).await.unwrap();
    assert_eq!(
        serde_json::to_value(&alone.entries).unwrap(),
        serde_json::to_value(&result.entries[9..]).unwrap()
    );
}
#[tokio::test]
async fn unavailable_discovery_is_not_a_matrix_request() {
    let transport = Synthetic {
        empty_discovery: true,
        ..Default::default()
    };
    let result = YtmClient::with_transport(transport.clone())
        .history(input(&["20260608", "20260609"], true))
        .await
        .unwrap();
    assert_eq!(
        (
            result.available_count,
            result.unavailable_count,
            result.data_row_count
        ),
        (0, 16, 0)
    );
    assert!(result.discovery.iter().all(|d| !d.available));
    assert_eq!(transport.requests.lock().unwrap().len(), 4);
    for entry in result.entries {
        let HistoryEntry::Unavailable {
            attempted_dates,
            stage,
            ..
        } = entry
        else {
            panic!("unavailable")
        };
        assert_eq!(attempted_dates.len(), 3);
        assert_eq!(serde_json::to_value(stage).unwrap(), "discovery");
    }
}
#[tokio::test]
async fn fatal_failure_stops_the_batch_with_context() {
    let transport = Synthetic {
        fatal: true,
        ..Default::default()
    };
    let error = YtmClient::with_transport(transport.clone())
        .history(input(&["20260608", "20260609"], false))
        .await
        .unwrap_err();
    assert_eq!(error.details.code, "source_protocol_error");
    assert_eq!(error.details.actual.as_ref().unwrap()["kind"]["code"], "20");
    assert_eq!(transport.requests.lock().unwrap().len(), 3);
    let client = YtmClient::with_transport(Synthetic {
        empty_discovery: true,
        ..Default::default()
    });
    let error = client
        .history(input(&["0000-01-01"], true))
        .await
        .unwrap_err();
    assert_eq!(error.details.code, "invalid_parameter");
}
#[tokio::test]
async fn cancellation_stops_inflight_history_and_next_call_is_independent() {
    let transport = Synthetic {
        wait: true,
        ..Default::default()
    };
    let client = YtmClient::with_transport(transport.clone());
    let token = CancellationToken::new();
    let request = client.history_with_cancellation(input(&["20260608"], false), token.clone());
    let cancel = async {
        while transport.requests.lock().unwrap().is_empty() {
            tokio::task::yield_now().await;
        }
        token.cancel();
    };
    let (result, ()) = tokio::join!(request, cancel);
    assert_eq!(
        result.unwrap_err().details.cause.as_deref(),
        Some("AbortError")
    );
    assert_eq!(transport.requests.lock().unwrap().len(), 1);
    let token = CancellationToken::new();
    token.cancel();
    assert!(client
        .history_with_cancellation(input(&["20260608"], false), token)
        .await
        .is_err());
    assert_eq!(transport.requests.lock().unwrap().len(), 1);
}

#[test]
fn full_date_limit_is_constructible() {
    let dates = DateSelection::range(
        BaseDate::new(2020, 1, 1).unwrap(),
        BaseDate::new(2025, 6, 22).unwrap(),
    )
    .unwrap();
    assert_eq!(dates.as_dates().len(), 2000);
}

fn count_input(count: usize, end: &str, start: Option<&str>) -> HistoryInput {
    HistoryInput::new(
        ytm_core::CountSelection::new(
            count,
            end.parse().unwrap(),
            start.map(|d| d.parse().unwrap()),
        )
        .unwrap(),
    )
}

#[tokio::test]
async fn count_180_stops_at_target_and_orders_whole_dates() {
    let transport = Synthetic::default();
    let requests = transport.requests.clone();
    let result = YtmClient::with_transport(transport)
        .history(count_input(180, "2026-06-29", None))
        .await
        .unwrap();
    assert_eq!(result.requested_dates.len(), 180);
    assert_eq!(result.requested_dates[0].to_string(), "2026-01-01");
    assert_eq!(result.requested_dates[179].to_string(), "2026-06-29");
    assert!(result.requested_dates.windows(2).all(|d| d[0] < d[1]));
    assert_eq!(requests.lock().unwrap().len(), 180 * 10);
    assert_eq!(result.available_count, 180 * 9);
    for day in result.entries.chunks(9) {
        assert_eq!(
            day.iter()
                .map(|e| match e {
                    HistoryEntry::Available { matrix } => {
                        assert!(!matrix.date_resolution.used_fallback);
                        assert_eq!(matrix.base_date, matrix.requested_base_date);
                        matrix.kind.code.as_str()
                    }
                    _ => panic!("numeric fixture"),
                })
                .collect::<Vec<_>>(),
            ["10", "20", "30", "40", "50", "60", "70", "80", "90"]
        );
    }
    assert_eq!(result.count_selection.unwrap().scanned_date_count, 180);
}

#[tokio::test]
async fn count_shortfalls_report_boundaries_and_preserve_fatal_errors() {
    for (end, start, scanned, stop) in [
        ("2024-03-01", Some("2024-02-28"), 3, "start_boundary"),
        ("2026-06-08", None, 2000, "search_limit"),
        ("0000-01-02", None, 2, "date_floor"),
    ] {
        let transport = Synthetic {
            empty_discovery: true,
            ..Default::default()
        };
        let requests = transport.requests.clone();
        let error = YtmClient::with_transport(transport)
            .history(count_input(1, end, start))
            .await
            .unwrap_err();
        assert_eq!(error.details.code, "insufficient_history");
        let actual = error.details.actual.unwrap();
        assert_eq!(actual["foundCount"], 0);
        assert_eq!(actual["scannedDateCount"], scanned);
        assert_eq!(actual["stopReason"], stop);
        assert_eq!(requests.lock().unwrap().len(), scanned as usize);
    }
    let error = YtmClient::with_transport(Synthetic {
        fatal: true,
        ..Default::default()
    })
    .history(count_input(1, "2026-06-08", None))
    .await
    .unwrap_err();
    assert_eq!(error.details.code, "source_protocol_error");
    assert_eq!(error.details.actual.unwrap()["kind"]["code"], "20");
}

#[test]
fn count_validation_is_exact_and_bounded() {
    for value in [
        serde_json::json!({"count":1}),
        serde_json::json!({"count":0,"endDate":"20260608"}),
        serde_json::json!({"count":2001,"endDate":"20260608"}),
        serde_json::json!({"count":1,"endDate":"20260608","baseDates":["20260608"]}),
        serde_json::json!({"count":1,"endDate":"20260608","fallback":"previous-available"}),
        serde_json::json!({"count":1,"endDate":"20260608","lookbackDays":1}),
        serde_json::json!({"count":1,"endDate":"20260608","startDate":"20260609"}),
        serde_json::json!({"count":1,"endDate":"20260608","startDate":"20200101"}),
    ] {
        assert!(
            HistoryInput::try_from(serde_json::from_value::<HistoryRequest>(value).unwrap())
                .is_err()
        );
    }
    for count in [
        serde_json::json!(true),
        serde_json::json!(1.5),
        serde_json::json!(-1),
        serde_json::json!("1"),
        serde_json::json!(1e40),
    ] {
        assert!(serde_json::from_value::<HistoryRequest>(
            serde_json::json!({"count":count,"endDate":"20260608"})
        )
        .is_err());
    }
}

#[tokio::test]
async fn count_accepts_the_last_candidate_and_rejects_rust_fallback_before_io() {
    let transport = Synthetic {
        fallback: true,
        ..Default::default()
    };
    let requests = transport.requests.clone();
    let client = YtmClient::with_transport(transport);
    let mut invalid = count_input(1, "2026-06-08", None);
    invalid.fallback = FallbackPolicy::PreviousAvailable(LookbackDays::default());
    assert_eq!(
        client.history(invalid).await.unwrap_err().details.code,
        "invalid_parameter"
    );
    assert!(requests.lock().unwrap().is_empty());
    let result = client
        .history(count_input(1, "2031-11-28", None))
        .await
        .unwrap();
    assert_eq!(result.requested_dates, ["2026-06-08".parse().unwrap()]);
    assert_eq!(result.count_selection.unwrap().scanned_date_count, 2000);
    assert_eq!(requests.lock().unwrap().len(), 2000 * 10);
}
